postgresql/postgresql_plugin.cc

```diff
@@ -0,0 +1,832 @@
+    {
+        auto conn = startConnection();
+
+        for (auto& vals : rows) {
+            insertRowPostgresql(conn, vals.second);
+        }        
+
+        PQfinish(conn);
+    }
+
+    /** Commit changes to the database.  Default is a no-op. */
+    virtual void commit() override
+    {
+    }
+
+    virtual std::pair<Date, Date> getTimestampRange() const override
+    {
+        throw HttpReturnException(400, "Postgresql recorder dataset is record-only");
+    }
+
+    virtual Date quantizeTimestamp(Date timestamp) const override
+    {
+        throw HttpReturnException(400, "Postgresql recorder dataset is record-only");
+    }
+
+    virtual std::shared_ptr<MatrixView> getMatrixView() const override
+    {
+        throw HttpReturnException(400, "Postgresql recorder dataset is record-only");
+    }
+
+    virtual std::shared_ptr<ColumnIndex> getColumnIndex() const override
+    {
+        throw HttpReturnException(400, "Postgresql recorder dataset is record-only");
+    }
+
+    virtual std::shared_ptr<RowStream> getRowStream() const override
+    {
+        throw HttpReturnException(400, "Postgresql recorder dataset is record-only");
+    }    
+
+};
+
+
+/*****************************************************************************/
+/* POSTGRESQL IMPORT PROCEDURE                                               */
+/*****************************************************************************/
+
+struct PostgresqlImportConfig: public ProcedureConfig {
+
+    static constexpr const char * name = "postgresql.import";
+
+    string databaseName;
+    int port;
+    string host;
+    string postgresqlQuery;
+
+    /// The output dataset.  Rows will be dumped into here via insertRows.
+    PolyConfigT<Dataset> outputDataset;
+
+    PostgresqlImportConfig() {
+        databaseName = "database";
+        port = 1234;
+        host = "localhost";
+        postgresqlQuery = "";
+
+        outputDataset.withType("sparse.mutable");
+    }
+
+};
+
+DECLARE_STRUCTURE_DESCRIPTION(PostgresqlImportConfig);
+DEFINE_STRUCTURE_DESCRIPTION(PostgresqlImportConfig);
+
+PostgresqlImportConfigDescription::
+PostgresqlImportConfigDescription()
+{
+    addParent<ProcedureConfig>();
+
+    addField("databaseName", &PostgresqlImportConfig::databaseName, "Name of the database to connect to.");
+    addField("port", &PostgresqlImportConfig::port, "Port of the database to connect to.", 1234);
+    addField("host", &PostgresqlImportConfig::host, "Address of the database to connect to ");
+    addField("postgresqlQuery", &PostgresqlImportConfig::postgresqlQuery, "Query to run in postgresql to get rows");
+
+    addField("outputDataset", &PostgresqlImportConfig::outputDataset,
+             "Output dataset configuration.  This may refer either to an "
+             "existing dataset, or a fully specified but non-existing dataset "
+             "which will be created by the procedure.", PolyConfigT<Dataset>().withType("sparse.mutable"));
```

Comment:
Consider adding validation of the database name here to make sure it is not empty and follow any other constraints forced by PostgreSQL.